crates/rpc/src/jsonrpc/websocket/logic.rs

```diff
@@ -0,0 +1,500 @@
+//! See [the parent module documentation](super)
+use std::collections::HashMap;
+use std::ops::ControlFlow;
+
+use crate::jsonrpc::request::RawParams;
+use crate::jsonrpc::{RequestId, RpcRequest};
+use axum::extract::ws::{Message, WebSocket};
+use axum::extract::{State, WebSocketUpgrade};
+use axum::response::IntoResponse;
+use futures::sink::Buffer;
+use futures::stream::{SplitSink, SplitStream};
+use futures::{SinkExt, StreamExt};
+use tokio::sync::mpsc;
+use tracing::error;
+
+use crate::jsonrpc::websocket::data::{Kind, ResponseEvent, SubscriptionId, SubscriptionItem};
+use crate::BlockHeader;
+
+const SUBSCRIBE_METHOD: &str = "pathfinder_subscribe";
+const UNSUBSCRIBE_METHOD: &str = "pathfinder_unsubscribe";
+const NEW_HEADS_TOPIC: &str = "newHeads";
+
+pub async fn websocket_handler(
+    ws: WebSocketUpgrade,
+    State(senders): State<WebsocketSenders>,
+) -> impl IntoResponse {
+    ws.on_upgrade(|socket| handle_socket(socket, senders))
+}
+
+async fn handle_socket(socket: WebSocket, senders: WebsocketSenders) {
+    let (ws_sender, ws_receiver) = socket.split();
+
+    let (msg_sender, msg_receiver) = mpsc::channel(100);
```

Comment:
If I understand correctly, this buffer capacity doesn't matter much; since the inner `Buffer` in `send_response` would be the first to fill-up, and then would break before this one really has a chance to?

Reply:
I've made the buffer sizes configurable.
Unfortunately I also rebased the PR in the same commit so it's not the easiest to review, but looking at just the `config.rs` in the [rebase commit](https://github.com/eqlabs/pathfinder/pull/1419/files#diff-04da37bcdbbc322d7ec6920e1971bb3100dc764738121fe4cc30f1e71f1417e1) file should give you the essence of this change.

Dropped configuration keys:
- `rpc.websocket`: Enable RPC WebSocket transport
- `rpc.websocket.capacity`: Maximum number of websocket subscriptions per subscription type

New configuration keys:
- `rpc.websocket.enabled`: Enable RPC WebSocket transport at the "/ws" path
- `rpc.websocket.buffer-capacity`: The socket buffer for outbound messages
- `rpc.websocket.topic-capacity`: The topic sender capacity. The topic senders are upstream of socket buffers and common to all clients and subscriptions

Please let me know if I should document these somewhere else than just in the clap attributes.

Reply:
> Please let me know if I should document these somewhere else than just in the clap attributes.

Just there is fine.

I have some concerns about the topic capacity though. I imagine it looks something like this:
```
                /--------- websocket buffer
               /---------- websocket buffer
topic buffer  < ---------- websocket buffer
               \---------- websocket buffer
                \--------- websocket buffer
```
i.e. a fan-out. And my concern is: can a single slow websocket thereby cause all other websockets on that same topic to die? 

If yes, then that's something we need to address/fix; probably by having it be a try-send from topic to websocket -- and then the websocket should get closed as too slow if this fails.

If no, then probably the operator has no real way of configuring or guessing what a good value for this should be. So we should just pick a reasonable number?

Reply:
> And my concern is: can a single slow websocket thereby cause all other websockets on that same topic to die?

The lagging `Receiver` will return an error on `recv` and the subscription will be dropped. If this happens too often to too many clients, consider increasing the topic buffer. If this happens sporadically to a few clients, consider increasing the websocket buffer.

I'll add these details to the clap arg descriptions.

Relevant Tokio doc:
> If the [Receiver](https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Receiver.html) handle falls behind, once the channel is full, newly sent values will overwrite old values. At this point, a call to [recv](https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Receiver.html#method.recv) will return with Err(RecvError::Lagged) and the [Receiver](https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Receiver.html)’s internal cursor is updated to point to the oldest value still held by the channel. A subsequent call to [recv](https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Receiver.html#method.recv) will return this value unless it has been since overwritten.

Reply:
Precision, I'll make adjustments so the client receives an error specifically mentioning a lagging stream and make the associated log record an `info` one.